dj/api/annotations.py

```diff
@@ -0,0 +1,127 @@
+"""
+Annotation related APIs.
+"""
+
+import logging
+from typing import List
+
+from fastapi import APIRouter, Depends
+from sqlmodel import Session, select
+
+from dj.errors import DJAlreadyExistsException, DJException
+from dj.models.annotation import (
+    RESERVED_ANNOTATION_NAMESPACE,
+    AnnotationType,
+    AnnotationTypeBase,
+)
+from dj.models.node import NodeType
+from dj.utils import get_session
+
+_logger = logging.getLogger(__name__)
+router = APIRouter()
+
+
+@router.get("/annotations/", response_model=List[AnnotationType])
+def list_annotations(
+    *, session: Session = Depends(get_session)
+) -> List[AnnotationType]:
+    """
+    List all available annotations.
+    """
+    annotation_types = session.exec(select(AnnotationType)).all()
+    return annotation_types
+
+
+@router.post("/annotations/", response_model=AnnotationTypeBase, status_code=201)
+def add_annotation_type(
+    data: AnnotationTypeBase, *, session: Session = Depends(get_session)
+) -> AnnotationType:
+    """
+    Add a new annotation type
+    """
+    if data.namespace == RESERVED_ANNOTATION_NAMESPACE:
+        raise DJException(
+            message="Cannot use `system` as the annotation type namespace as it is reserved.",
+        )
+    statement = select(AnnotationType).where(AnnotationType.name == data.name)
+    annotation_type = session.exec(statement).unique().one_or_none()
+    if annotation_type:
+        raise DJAlreadyExistsException(
+            message=f"Annotation type `{data.name}` already exists!",
+        )
+    annotation_type = AnnotationType.from_orm(data)
+    session.add(annotation_type)
+    session.commit()
+    session.refresh(annotation_type)
+    return annotation_type
+
+
+def default_annotation_types(session: Session = Depends(get_session)):
+    """
+    Loads all the column annotation types that are supported by the system
+    by default into the database.
+    """
+    defaults = [
+        AnnotationType(
+            namespace=RESERVED_ANNOTATION_NAMESPACE,
+            name="primary_key",
+            description="Points to a column which is part of the primary key of the node",
```

Comment:
This just made me thought of partition keys. And the amazing thing is that you could almost implement them using the same mechanism, except the fact that list of partition key columns should be ordered... so should we sneak in a `position` attribute to our ColumnAnnotation/ColumnAttribute class and have a solved problem. Or should we keep partition attributes separate? I am leaning towards the latter, but I wanted to ask you all.

Reply:
Yeah, this is a good idea! It makes sense to me that partitions are just attributes on columns. Although I think this may be better off in a separate PR so that we can also add in more logic around enforcement of partition positions.